File: cpp/gestores/gestor_academico.cpp
#include "gestor_academico.h"
#include <ctime>
#include <iomanip>
#include <sstream>

namespace pita {

static std::string fecha_hoy() {
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%d");
    return oss.str();
}

GestorAcademico::GestorAcademico(
    ListaEnlazada<PlanEstudio>& planes,
    ListaEnlazada<DetallePlanEstudio>& detallesPlanes,
    ListaEnlazada<Curso>& cursos,
    ListaEnlazada<Prerrequisito>& prerrequisitos,
    ListaEnlazada<ProgramaAcademico>& programas,
    ListaEnlazada<Horario>& horarios,
    ListaEnlazada<AsignacionDocente>& asignaciones,
    ListaEnlazada<Profesor>& profesores,
    ListaEnlazada<OfertaCurso>& ofertas
) : planes(planes),
    detallesPlanes(detallesPlanes),
    cursos(cursos),
    prerrequisitos(prerrequisitos),
    programas(programas),
    horarios(horarios),
    asignaciones(asignaciones),
    profesores(profesores),
    ofertas(ofertas) {}

int GestorAcademico::siguienteIdPlan() {
    int maxId = 0;
    for (const auto& p : planes) {
        if (p.idPlanEstudio.has_value() && *p.idPlanEstudio > maxId) {
            maxId = *p.idPlanEstudio;
        }
    }
    return maxId + 1;
}

int GestorAcademico::siguienteIdDetallePlan() {
    int maxId = 0;
    for (const auto& d : detallesPlanes) {
        if (d.idDetallePlan.has_value() && *d.idDetallePlan > maxId) {
            maxId = *d.idDetallePlan;
        }
    }
    return maxId + 1;
}

int GestorAcademico::siguienteIdPrerrequisito() {
    int maxId = 0;
    for (const auto& p : prerrequisitos) {
        if (p.idPrerrequisito.has_value() && *p.idPrerrequisito > maxId) {
            maxId = *p.idPrerrequisito;
        }
    }
    return maxId + 1;
}

int GestorAcademico::siguienteIdAsignacion() {
    int maxId = 0;
    for (const auto& a : asignaciones) {
        if (a.idAsignacion.has_value() && *a.idAsignacion > maxId) {
            maxId = *a.idAsignacion;
        }
    }
    return maxId + 1;
}

int GestorAcademico::siguienteIdHorario() {
    int maxId = 0;
    for (const auto& h : horarios) {
        if (h.idHorario.has_value() && *h.idHorario > maxId) {
            maxId = *h.idHorario;
        }
    }
    return maxId + 1;
}

bool GestorAcademico::seCruzan(const Horario& primero, const Horario& segundo) {
    if (!primero.diaSemana.has_value() || !segundo.diaSemana.has_value()) return false;
    if (*primero.diaSemana != *segundo.diaSemana) return false;
    if (!primero.horaInicio.has_value() || !primero.horaFin.has_value()) return false;
    if (!segundo.horaInicio.has_value() || !segundo.horaFin.has_value()) return false;
    return (*primero.horaInicio < *segundo.horaFin && *segundo.horaInicio < *primero.horaFin);
}

PlanEstudio& GestorAcademico::crearPlan(PlanEstudio plan) {
    if (!plan.idPlanEstudio.has_value()) {
        plan.idPlanEstudio = siguienteIdPlan();
    } else {
        for (const auto& p : planes) {
            if (p.idPlanEstudio.has_value() && *p.idPlanEstudio == *plan.idPlanEstudio) {
                throw ErrorAcademico("Ya existe un plan de estudio con ID " + std::to_string(*plan.idPlanEstudio));
            }
        }
    }

    if (plan.idPrograma.has_value()) {
        bool existeProg = false;
        for (const auto& pr : programas) {
            if (pr.idPrograma.has_value() && *pr.idPrograma == *plan.idPrograma) {
                existeProg = true;
                break;
            }
        }
        if (!existeProg) {
            throw ErrorAcademico("No existe el programa con ID " + std::to_string(*plan.idPrograma));
        }
    }

    if (plan.fechaFinVigencia.has_value() && plan.fechaInicioVigencia.has_value()) {
        if (*plan.fechaFinVigencia < *plan.fechaInicioVigencia) {
            throw ErrorAcademico("La vigencia final del plan no puede preceder a la inicial");
        }
    }

    if (!plan.estado.has_value() || plan.estado->empty()) {
        plan.estado = "ACTIVO";
    }

    planes.push_back(std::move(plan));
    return planes.back();
}

DetallePlanEstudio& GestorAcademico::incluirCurso(int idPlan, int idCurso, int semestreSugerido, bool esObligatorio) {
    PlanEstudio* plan = nullptr;
    for (auto& p : planes) {
        if (p.idPlanEstudio.has_value() && *p.idPlanEstudio == idPlan) {
            plan = &p;
            break;
        }
    }
    if (!plan) {
        throw ErrorAcademico("No existe el plan de estudio con ID " + std::to_string(idPlan));
    }

    Curso* curso = nullptr;
    for (auto& c : cursos) {
        if (c.idCurso.has_value() && *c.idCurso == idCurso) {
            curso = &c;
            break;
        }
    }
    if (!curso) {
        throw ErrorAcademico("No existe el curso con ID " + std::to_string(idCurso));
    }

    if (semestreSugerido < 1) {
        throw ErrorAcademico("El semestre sugerido debe ser positivo");
    }

    for (const auto& d : detallesPlanes) {
        if (d.idPlanEstudio.has_value() && *d.idPlanEstudio == idPlan &&
            d.idCurso.has_value() && *d.idCurso == idCurso) {
            throw ErrorAcademico("El curso ya pertenece a este plan");
        }
    }

    DetallePlanEstudio detalle;
    detalle.idDetallePlan = siguienteIdDetallePlan();
    detalle.idPlanEstudio = idPlan;
    detalle.idCurso = idCurso;
    detalle.semestreSugerido = semestreSugerido;
    detalle.numeroCreditos = curso->numeroCreditos;
    detalle.esObligatorio = esObligatorio;
    detalle.estado = "ACTIVO";

    detallesPlanes.push_back(std::move(detalle));

    int totalCred = 0;
    for (const auto& d : detallesPlanes) {
        if (d.idPlanEstudio.has_value() && *d.idPlanEstudio == idPlan &&
            (!d.estado.has_value() || *d.estado != "INACTIVO")) {
            totalCred += d.numeroCreditos.value_or(0);
        }
    }
    plan->totalCreditos = totalCred;

    return detallesPlanes.back();
}

Prerrequisito& GestorAcademico::registrarPrerrequisito(Prerrequisito prerrequisito) {
    bool cursoExiste = false;
    bool reqExiste = false;
    for (const auto& c : cursos) {
        if (c.idCurso.has_value()) {
            if (prerrequisito.idCurso.has_value() && *c.idCurso == *prerrequisito.idCurso) cursoExiste = true;
            if (prerrequisito.idCursoRequerido.has_value() && *c.idCurso == *prerrequisito.idCursoRequerido) reqExiste = true;
        }
    }
    if (!cursoExiste) {
        throw ErrorAcademico("No existe el curso con ID " + (prerrequisito.idCurso.has_value() ? std::to_string(*prerrequisito.idCurso) : ""));
    }
    if (!reqExiste) {
        throw ErrorAcademico("No existe el curso prerrequisito con ID " + (prerrequisito.idCursoRequerido.has_value() ? std::to_string(*prerrequisito.idCursoRequerido) : ""));
    }

    if (prerrequisito.idCurso == prerrequisito.idCursoRequerido) {
        throw ErrorAcademico("Un curso no puede ser prerrequisito de si mismo");
    }

    for (const auto& p : prerrequisitos) {
        if (p.idCurso == prerrequisito.idCurso && p.idCursoRequerido == prerrequisito.idCursoRequerido) {
            throw ErrorAcademico("El prerrequisito ya esta registrado");
        }
    }

    if (!prerrequisito.idPrerrequisito.has_value()) {
        prerrequisito.idPrerrequisito = siguienteIdPrerrequisito();
    } else {
        for (const auto& p : prerrequisitos) {
            if (p.idPrerrequisito.has_value() && *p.idPrerrequisito == *prerrequisito.idPrerrequisito) {
                throw ErrorAcademico("Ya existe un prerrequisito con ID " + std::to_string(*prerrequisito.idPrerrequisito));
            }
        }
    }

    if (!prerrequisito.estado.has_value() || prerrequisito.estado->empty()) {
        prerrequisito.estado = "ACTIVO";
    }

    prerrequisitos.push_back(std::move(prerrequisito));
    return prerrequisitos.back();
}

std::pair<PlanEstudio*, ListaEnlazada<DetallePlanEstudio>> GestorAcademico::consultarPlan(int idPlan) {
    PlanEstudio* plan = nullptr;
    for (auto& p : planes) {
        if (p.idPlanEstudio.has_value() && *p.idPlanEstudio == idPlan) {
            plan = &p;
            break;
        }
    }
    if (!plan) {
        throw ErrorAcademico("No existe el plan de estudio con ID " + std::to_string(idPlan));
    }

    ListaEnlazada<DetallePlanEstudio> detalles;
    for (const auto& d : detallesPlanes) {
        if (d.idPlanEstudio.has_value() && *d.idPlanEstudio == idPlan) {
            detalles.push_back(d);
        }
    }
    return {plan, detalles};
}

AsignacionDocente& GestorAcademico::asignarProfesor(int idProfesor, int idOferta, double numeroHoras) {
    bool profExiste = false;
    for (const auto& p : profesores) {
        if (p.idProfesor.has_value() && *p.idProfesor == idProfesor) {
            profExiste = true;
            break;
        }
    }
    if (!profExiste) {
        throw ErrorAcademico("No existe el profesor con ID " + std::to_string(idProfesor));
    }

    bool ofertaExiste = false;
    for (const auto& o : ofertas) {
        if (o.idOfertaCurso.has_value() && *o.idOfertaCurso == idOferta) {
            ofertaExiste = true;
            break;
        }
    }
    if (!ofertaExiste) {
        throw ErrorAcademico("No existe la oferta con ID " + std::to_string(idOferta));
    }

    for (const auto& a : asignaciones) {
        if (a.idProfesor.has_value() && *a.idProfesor == idProfesor &&
            a.idOfertaCurso.has_value() && *a.idOfertaCurso == idOferta) {
            throw ErrorAcademico("El profesor ya esta asignado a esta oferta");
        }
    }

    if (numeroHoras <= 0) {
        throw ErrorAcademico("El numero de horas debe ser positivo");
    }

    AsignacionDocente asig;
    asig.idAsignacion = siguienteIdAsignacion();
    asig.idProfesor = idProfesor;
    asig.idOfertaCurso = idOferta;
    asig.numeroHoras = numeroHoras;
    asig.fechaAsignacion = fecha_hoy();
    asig.estado = "ACTIVO";

    asignaciones.push_back(std::move(asig));
    return asignaciones.back();
}

Horario& GestorAcademico::agregarHorario(Horario horario) {
    if (!horario.horaInicio.has_value() || !horario.horaFin.has_value() || *horario.horaInicio >= *horario.horaFin) {
        throw ErrorAcademico("El horario debe tener horas validas y consecutivas");
    }

    for (const auto& existente : horarios) {
        if (existente.idOfertaCurso.has_value() && horario.idOfertaCurso.has_value() &&
            *existente.idOfertaCurso == *horario.idOfertaCurso) {
            if (seCruzan(horario, existente)) {
                throw ErrorAcademico("El horario se cruza con otro horario de la oferta");
            }
        }
    }

    if (!horario.idHorario.has_value()) {
        horario.idHorario = siguienteIdHorario();
    } else {
        for (const auto& h : horarios) {
            if (h.idHorario.has_value() && *h.idHorario == *horario.idHorario) {
                throw ErrorAcademico("Ya existe un horario con ID " + std::to_string(*horario.idHorario));
            }
        }
    }

    horarios.push_back(std::move(horario));
    return horarios.back();
}

PlanEstudio& GestorAcademico::desactivarPlan(int idPlan) {
    for (auto& p : planes) {
        if (p.idPlanEstudio.has_value() && *p.idPlanEstudio == idPlan) {
            p.estado = "INACTIVO";
            return p;
        }
    }
    throw ErrorAcademico("No existe el plan de estudio con ID " + std::to_string(idPlan));
}

} // namespace pita

File: gestores/gestor_academico.py
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Any, Iterable

from modelo_datos import AsignacionDocente, Curso, DetallePlanEstudio, Horario, PlanEstudio, Prerrequisito


class ErrorAcademico(ValueError):
    """Error de estructura o referencia académica."""


class GestorAcademico:
    def __init__(
        self,
        planes: list[PlanEstudio],
        detalles_planes: list[DetallePlanEstudio],
        cursos: list[Curso],
        prerrequisitos: list[Prerrequisito],
        programas: Iterable[Any] | None = None,
        horarios: list[Horario] | None = None,
        asignaciones: list[AsignacionDocente] | None = None,
        profesores: Iterable[Any] | None = None,
        ofertas: Iterable[Any] | None = None,
    ) -> None:
        self.planes = planes
        self.detalles_planes = detalles_planes
        self.cursos = cursos
        self.prerrequisitos = prerrequisitos
        self.programas = list(programas or [])
        self.horarios = horarios if horarios is not None else []
        self.asignaciones = asignaciones if asignaciones is not None else []
        self.profesores = list(profesores or [])
        self.ofertas = list(ofertas or [])

    def crear_plan(self, plan: PlanEstudio) -> PlanEstudio:
        if plan.idPlanEstudio is None:
            plan.idPlanEstudio = self._siguiente_id(self.planes, "idPlanEstudio")
        self._id_unico(self.planes, "idPlanEstudio", plan.idPlanEstudio, "plan de estudio")
        if plan.idPrograma is not None:
            self._referencia(self.programas, "idPrograma", plan.idPrograma, "programa")
        if plan.fechaFinVigencia and plan.fechaInicioVigencia and plan.fechaFinVigencia < plan.fechaInicioVigencia:
            raise ErrorAcademico("La vigencia final del plan no puede preceder a la inicial")
        plan.estado = plan.estado or "ACTIVO"
        self.planes.append(plan)
        return plan

    def incluir_curso(
        self,
        id_plan: int,
        id_curso: int,
        semestre_sugerido: int,
        *,
        es_obligatorio: bool = True,
    ) -> DetallePlanEstudio:
        plan = self._buscar(self.planes, "idPlanEstudio", id_plan, "plan de estudio")
        curso = self._buscar(self.cursos, "idCurso", id_curso, "curso")
        if semestre_sugerido < 1:
            raise ErrorAcademico("El semestre sugerido debe ser positivo")
        if any(item.idPlanEstudio == id_plan and item.idCurso == id_curso for item in self.detalles_planes):
            raise ErrorAcademico("El curso ya pertenece a este plan")
        detalle = DetallePlanEstudio(
            idDetallePlan=self._siguiente_id(self.detalles_planes, "idDetallePlan"),
            idPlanEstudio=plan.idPlanEstudio,
            idCurso=curso.idCurso,
            semestreSugerido=semestre_sugerido,
            numeroCreditos=curso.numeroCreditos,
            esObligatorio=es_obligatorio,
            estado="ACTIVO",
        )
        self.detalles_planes.append(detalle)
        plan.totalCreditos = sum(
            item.numeroCreditos or 0
            for item in self.detalles_planes
            if item.idPlanEstudio == id_plan and item.estado != "INACTIVO"
        )
        return detalle

    def registrar_prerrequisito(self, prerrequisito: Prerrequisito) -> Prerrequisito:
        self._buscar(self.cursos, "idCurso", prerrequisito.idCurso, "curso")
        self._buscar(self.cursos, "idCurso", prerrequisito.idCursoRequerido, "curso prerrequisito")
        if prerrequisito.idCurso == prerrequisito.idCursoRequerido:
            raise ErrorAcademico("Un curso no puede ser prerrequisito de sí mismo")
        if any(
            item.idCurso == prerrequisito.idCurso
            and item.idCursoRequerido == prerrequisito.idCursoRequerido
            for item in self.prerrequisitos
        ):
            raise ErrorAcademico("El prerrequisito ya está registrado")
        if prerrequisito.idPrerrequisito is None:
            prerrequisito.idPrerrequisito = self._siguiente_id(self.prerrequisitos, "idPrerrequisito")
        self._id_unico(self.prerrequisitos, "idPrerrequisito", prerrequisito.idPrerrequisito, "prerrequisito")
        prerrequisito.estado = prerrequisito.estado or "ACTIVO"
        self.prerrequisitos.append(prerrequisito)
        return prerrequisito

    def consultar_plan(self, id_plan: int) -> tuple[PlanEstudio, list[DetallePlanEstudio]]:
        plan = self._buscar(self.planes, "idPlanEstudio", id_plan, "plan de estudio")
        detalles = [item for item in self.detalles_planes if item.idPlanEstudio == id_plan]
        return plan, detalles

    def asignar_profesor(self, id_profesor: int, id_oferta: int, numero_horas: Decimal) -> AsignacionDocente:
        self._referencia(self.profesores, "idProfesor", id_profesor, "profesor")
        self._referencia(self.ofertas, "idOfertaCurso", id_oferta, "oferta")
        if any(item.idProfesor == id_profesor and item.idOfertaCurso == id_oferta for item in self.asignaciones):
            raise ErrorAcademico("El profesor ya está asignado a esta oferta")
        if numero_horas <= 0:
            raise ErrorAcademico("El número de horas debe ser positivo")
        asignacion = AsignacionDocente(
            idAsignacion=self._siguiente_id(self.asignaciones, "idAsignacion"),
            idProfesor=id_profesor,
            idOfertaCurso=id_oferta,
            numeroHoras=numero_horas,
            fechaAsignacion=date.today(),
            estado="ACTIVO",
        )
        self.asignaciones.append(asignacion)
        return asignacion

    def agregar_horario(self, horario: Horario) -> Horario:
        if horario.horaInicio is None or horario.horaFin is None or horario.horaInicio >= horario.horaFin:
            raise ErrorAcademico("El horario debe tener horas válidas y consecutivas")
        existentes = [item for item in self.horarios if item.idOfertaCurso == horario.idOfertaCurso]
        if any(self._se_cruzan(horario, existente) for existente in existentes):
            raise ErrorAcademico("El horario se cruza con otro horario de la oferta")
        horario.idHorario = horario.idHorario or self._siguiente_id(self.horarios, "idHorario")
        if any(item.idHorario == horario.idHorario for item in self.horarios):
            raise ErrorAcademico(f"Ya existe un horario con ID {horario.idHorario}")
        self.horarios.append(horario)
        return horario

    @staticmethod
    def _se_cruzan(primero: Horario, segundo: Horario) -> bool:
        return primero.diaSemana == segundo.diaSemana and primero.horaInicio < segundo.horaFin and segundo.horaInicio < primero.horaFin

    def desactivar_plan(self, id_plan: int) -> PlanEstudio:
        plan = self._buscar(self.planes, "idPlanEstudio", id_plan, "plan de estudio")
        plan.estado = "INACTIVO"
        return plan

    @staticmethod
    def _buscar(elementos: Iterable[Any], campo: str, valor: Any, nombre: str) -> Any:
        elemento = next((item for item in elementos if getattr(item, campo, None) == valor), None)
        if elemento is None:
            raise ErrorAcademico(f"No existe el {nombre} con ID {valor}")
        return elemento

    @staticmethod
    def _referencia(elementos: Iterable[Any], campo: str, valor: Any, nombre: str) -> None:
        if not any(getattr(item, campo, None) == valor for item in elementos):
            raise ErrorAcademico(f"No existe el {nombre} con ID {valor}")

    @staticmethod
    def _id_unico(elementos: Iterable[Any], campo: str, valor: Any, nombre: str) -> None:
        if any(getattr(item, campo, None) == valor for item in elementos):
            raise ErrorAcademico(f"Ya existe un {nombre} con ID {valor}")

    @staticmethod
    def _siguiente_id(elementos: Iterable[Any], campo: str) -> int:
        return max((getattr(item, campo) or 0 for item in elementos), default=0) + 1
